Css grid

przytu1 opened the pull request:
Short description (with a reference to an issue):
https://github.com/Appsilon/shiny.semantic/issues/222

**DoD**

- [x] Major project work has a corresponding task. If there’s no task for what you are doing, create it. Each task needs to be well defined and described.

- [x] Change has been tested (manually or with automated tests), everything runs correctly and works as expected. No existing functionality is broken.

- [x] No new error or warning messages are introduced.

- [x] All interaction with a semantic functions, examples and docs are written from the perspective of the person using or receiving it. They are understandable and helpful to this person.

- [x] If the change affects code or repo sctructure, README, documentation and code comments should be updated. 

- [x] All code has been peer-reviewed before merging into any main branch.

- [x] All changes have been merged into the main branch we use for development (develop).

- [x] Continuous integration checks (linter, unit tests) are configured and passed.

- [x] Unit tests added for all new or changed logic.

- [ ] All task requirements satisfied. The reviewer is responsible to verify each aspect of the task.

- [ ] Any added or touched code follows our style-guide.

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+#' Generate CSS string representing grid template areas.
+#' @param areas_dataframe data.frame of strings representing grid areas
+#' @return string
```
I think that text format is called `character` in R

dokato replied:
I noticed there's more of these, could you normalise that across the script?

przytu1 replied:
done :heavy_check_mark:

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+#'
+#' @param area_names
+#'
+#' @return list of shiny::tags$div
```
could you wrap `shiny::tags$div` in `\code`?

przytu1 replied:
done :heavy_check_mark:

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+#'   display_grid(myGrid)
+#' }
+#' Second example:
+#' \preformatted{
```
For not exported function using `\preformatted` for code chunks is actually better but for the exported, we typically use `\examples`. Could you replace it?

przytu1 replied:
Done for `grid` and `grid_template` :heavy_check_mark:

dokato replied:
That's a bit weird, look how we do it elsewhere: https://github.com/Appsilon/shiny.semantic/blob/develop/R/rating.R#L15 and it works fine. Anyway, this is a minor thing. As long as is in correct section it's good enough.

przytu1 replied:
Ok, thanks for this link! I fixed it again, now it works with `@examples` only

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+#' @export
+#'
+grid_template <- function(
+  default = list(areas = "", rows_height = c(), cols_width = c()),
```
are these default parameters any useful here, or they're just placeholders for people to fill in? if the latter, I think they should be removed and just put in the description. `default` is mandatory to specify anyway, isn't it?

przytu1 replied:
I refactored it to more concrete argument, so it is indeed a default, not just a placeholder to fill in. Now when you call `grid_template()` you will get a grid with single area 100% x 100% size:

```
default = list(areas = rbind(c("main_area")), rows_height = c("100%"), cols_width = c("100%"))
```

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+#'
+grid_template <- function(
+  default = list(areas = "", rows_height = c(), cols_width = c()),
+  mobile = list(areas = "", rows_height = c(), cols_width = c())) {
```
Here I was thinking: how to make it most flexible, but at the same time user-friendly. We talked about it briefly and you mentioned yourself that there's a handful of other scenarios to consider, like `tablet` format for example... Maybe unnamed parameters would do (`...`), assuming that they have specified format.

Also, is there a plan to add in `grid_template` the conditions for switching between mobile and default?

przytu1 replied:
Use of ellipsis (`...`) here should be considered after we implement the logic that is responsible for switching between grids depending on the screen size. This is a good idea, but let's see how our logic works first.

In the simplest implementation this logic for now can decide when it displays mobile (hardcoded rules for screen sizes, "if width <800px then mobile"). Later we can allow user to define their own rules / media queries. But it has to be easy to learn and use for R users. I see at least two ideas:

* it can be an additional parameter on the `list(areas = ...)` object e.g. `max.width` and `min.width` 
* it can be argument e.g. `media.rules = list(mobile = "800px", tablet = "1200px")`

The logic for switching between mobile and default should use the following technique of combining media queries `@media` and css grid: https://www.w3.org/WAI/WCAG21/Techniques/css/C32.html

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+  custom_styles <- list(custom_style_grid_container = container_style)
+  for (area in area_names) {
+    custom_styles[[glue::glue("custom_style_grid_area_{area}")]] <- ifelse(
+      area %in% names(area_styles), area_styles[[area]], "")
```
I'm surprised that the checks have passed. I thought it would complain about lack of import magrittr here (`%in%`), but anyway if it works with `--as-cran` checks we may leave it as is

przytu1 replied:
`%in%` is base R operator for checking if vector contains an element, it's not imported from magrittr :)

dokato commented on R/grid.R:
```diff
@@ -0,0 +1,295 @@
+#'
+#' Example:
+#'
+#' \preformatted{
```
here again, `@examples` please

przytu1 replied:
done :heavy_check_mark: 

```
#' @examples
#' \preformatted{...}
```

przytu1 commented:
Thank you @dokato! I addressed all comments. I am also adding a backlog ticket for implementing grid examples: https://github.com/Appsilon/shiny.semantic/issues/240
